posthog/permissions.py

```diff
@@ -242,3 +246,106 @@ def has_permission(self, request, view) -> bool:
+
+        for required_scope in required_scopes:
+            valid_scopes = [required_scope]
+
+            # For all valid scopes with :read we also add :write
+            for scope in valid_scopes:
+                if scope.endswith(":read"):
+                    valid_scopes.append(scope.replace(":read", ":write"))
+
+            if not any(scope in key_scopes for scope in valid_scopes):
+                raise PermissionDenied(f"API key missing required scope '{required_scope}'")
+
+        return True
+
+    def check_team_and_org_permissions(self, request, view) -> None:
+        scoped_organizations = request.successful_authenticator.personal_api_key.scoped_organizations_list
+        scoped_teams = request.successful_authenticator.personal_api_key.scoped_teams_list
+
+        if scoped_organizations:
+            try:
+                organization = get_organization_from_view(view)
+            except ValueError:
+                # Indicates this is not an organization scoped view
+                pass
+            if str(organization.id) not in scoped_organizations:
+                raise PermissionDenied(
+                    f"API key does not have access to the requested organization '{organization.id}'"
+                )
+
+        if scoped_teams:
+            try:
+                team = view.team
+                if team.id not in scoped_teams:
+                    raise PermissionDenied(f"API key does not have access to the requested team '{team.id}'")
+
+            except (ValueError, KeyError):
+                # Indicates this is not a team scoped view
+                pass
+
+    def get_required_scopes(self, request, view) -> list[str]:
+        # If required_scopes is set on the view method then use that
+        # Otherwise use the base_scope and derive the required scope from the action
+        if getattr(view, "required_scopes", None):
+            return view.required_scopes
+
+        base_scope = self.get_base_scope(request, view)
+
+        if base_scope == "not_supported":
+            raise PermissionDenied(f"This action does not support Personal API Key access")
+
+        read_actions = getattr(view, "base_scope_read_actions", self.read_actions)
+        write_actions = getattr(view, "base_scope_write_actions", self.write_actions)
+
+        if view.action in write_actions:
+            return [f"{base_scope}:write"]
+        elif view.action in read_actions or request.method == "OPTIONS":
+            return [f"{base_scope}:read"]
+
+        # If we get here this typically means an action was called without a required scope
+        # It is essentially "not_supported"
+        raise PermissionDenied(f"This action does not support Personal API Key access")
+
+    def get_base_scope(self, request, view) -> APIScopeObjectOrNotSupported:
+        if not getattr(view, "base_scope", None):
+            raise ImproperlyConfigured("APIScopePermission requires the view to define the base_scope attribute.")
```

Comment:
If `base_scope` is 100% mandatory, then we should just do this check in `__init__()`

Reply:
I don't fully get that? Which __init__ ?

Reply:
`APIScopePermission`'s `__init__`. It would have just three lines: two for this `if not getattr(view, "base_scope", None):` check and one for `super().__init__(*args, **kwargs)`

Reply:
Ah, wait, I actually thought this is on the viewset. So when I wrote out "`APIScopePermission`'s" just now it clicked for me that we can't quite do that. This is alright then

Reply:
Where does it get `view` from though?